File: crates/bliti-web/app/src/readings.js
// Making sense of the facts and readings of NFO, and rendering them the way VIEW asks ours to.
//
// The wire carries data; the intelligence is here. This module holds our application's own order, its
// wording for the catalogue it recognises, and its aggregation rules. Everything it does not
// recognise still renders, from the entry's own name, kind and traits, so a device ahead of this
// build degrades to a plain reading card rather than vanishing (VIEW).

/// One fact or reading, normalised from a message. `fact` is true for a fact, false for a reading;
/// the two catalogues are separate and a name may appear in both.
export function entryOf(message) {
	return {
		fact: message.type === 'fact',
		name: message.fact ?? message.measurement,
		at: message.at,
		traits: message.traits ?? {},
		kind: message.kind,
		unit: message.unit,
		value: message.value,
	}
}

/// The status trait's `is`, or null. A status this build does not know is left as it arrived, and
/// treated as passed rather than crying wolf.
export function statusOf(entry) {
	return entry.traits?.status?.is ?? null
}

export function reasonOf(entry) {
	return entry.traits?.status?.reason ?? null
}

/// Whether a reading should draw an operator's attention. Only the two states with a notion of
/// difficulty do; skipped and broken say nothing is wrong with the thing measured, only that there is
/// no value.
export function isTrouble(entry) {
	return statusOf(entry) === 'warning' || statusOf(entry) === 'failed'
}

/// Whether the entry has a value at all. Skipped and broken carry none (NFO). Tolerates a missing
/// entry, so a fold that names one the device did not send is simply empty.
export function hasValue(entry) {
	return entry != null && entry.value !== undefined && entry.value !== null
}

// The order our application renders the catalogue it recognises in, and the wording it gives each
// name. The order is fixed and does not move when a reading goes into difficulty (VIEW).
export const HEADER = ['hostname', 'board', 'board-revision', 'os', 'kernel']

export const TILE_ORDER = [
	'network-address',
	'wireless-network',
	'hotspot',
	'hotspot-clients',
	'cpu-usage',
	'memory-usage',
	'filesystem-usage',
	'network-throughput',
	'temperature',
	'cpu-frequency',
	'fan-speed',
	'power-source',
	'battery-charge',
	'last-boot',
]

// Whether the network runs what was recorded or a proposal being tried. It marks the tiles below
// rather than getting one of its own (VIEW).
export const NETWORK_CONFIGURATION = 'network-configuration'
export const PROVISIONAL_TILES = new Set(['network-address', 'wireless-network', 'hotspot', 'hotspot-clients'])

// Entries rendered inside another's reveal rather than as a tile of their own.
export const IN_REVEAL = new Set([
	'memory-total',
	'filesystem-total',
	'cpu-frequency-max',
	'battery-voltage',
	'battery-direction',
])

// Our wording for each catalogue name. A name not here is title-cased from the name itself.
const LABELS = {
	'network-address': 'Address',
	'wireless-network': 'Wireless',
	hotspot: 'Hotspot',
	'hotspot-clients': 'Hotspot clients',
	'cpu-usage': 'Processor',
	'memory-usage': 'Memory',
	'filesystem-usage': 'Storage',
	'network-throughput': 'Network',
	temperature: 'Temperature',
	'cpu-frequency': 'Processor speed',
	'fan-speed': 'Fan',
	'power-source': 'Power',
	'battery-charge': 'Battery',
	'last-boot': 'Uptime',
}

// Our wording for the values of the catalogue's descriptive text readings.
const VALUES = {
	'power-source': { 'via-backup': 'On backup', battery: 'On battery', 'bypassing-backup': 'Backup bypassed' },
	'battery-direction': { charging: 'Charging', discharging: 'Discharging', idle: 'Idle' },
}

export function labelOf(name) {
	return LABELS[name] ?? titleCase(name)
}

function titleCase(name) {
	return name
		.split(/[-_]/)
		.map((word) => word.charAt(0).toUpperCase() + word.slice(1))
		.join(' ')
}

/// A value as an operator reads it: by its kind, in our own wording and at our own magnitude, or the
/// value stringified where the kind is one this build does not know (VIEW).
export function formatValue(entry) {
	if (!hasValue(entry)) return null
	const { kind, value, unit, name } = entry
	switch (kind) {
		case 'fraction':
			return `${Math.round(value * 100)}%`
		case 'quantity':
			return formatQuantity(value, unit)
		case 'duration':
			return formatDuration(value)
		case 'datetime':
			return name === 'last-boot' ? elapsedSince(value) : String(value)
		case 'text':
			return VALUES[name]?.[value] ?? String(value)
		case 'ipv4':
		case 'ipv6':
			return String(value)
		default:
			// A kind this build does not know: the value stringified, with the unit where there is one.
			return unit ? `${stringify(value)} ${unit}` : stringify(value)
	}
}

function stringify(value) {
	return typeof value === 'object' ? JSON.stringify(value) : String(value)
}

/// A quantity written in our own abbreviation and magnitude. An unrecognised unit is written out as
/// it was sent (VIEW).
function formatQuantity(number, unit) {
	switch (unit) {
		case 'bytes':
			return magnitude(number, 'B', 'kB', 'MB', 'GB')
		case 'bytes/second':
			return magnitude(number, 'B/s', 'kB/s', 'MB/s', 'GB/s')
		case 'hertz':
			return magnitude(number, 'Hz', 'kHz', 'MHz', 'GHz')
		case 'celsius':
			return `${trim(number)} °C`
		case 'volts':
			return `${trim(number)} V`
		case 'revolutions/minute':
			return `${Math.round(number)} rpm`
		case 'clients':
			return number === 1 ? '1 client' : `${trim(number)} clients`
		default:
			return `${trim(number)} ${unit}`
	}
}

function magnitude(number, ...units) {
	let value = number
	let index = 0
	while (value >= 1000 && index < units.length - 1) {
		value /= 1000
		index += 1
	}
	return `${trim(Math.round(value * 100) / 100)} ${units[index]}`
}

function trim(number) {
	return Number.isInteger(number) ? String(number) : String(Math.round(number * 100) / 100)
}

function formatDuration(seconds) {
	const whole = Math.floor(seconds)
	const days = Math.floor(whole / 86400)
	const hours = Math.floor((whole % 86400) / 3600)
	const minutes = Math.floor((whole % 3600) / 60)
	if (days > 0) return `${days}d ${hours}h`
	if (hours > 0) return `${hours}h ${minutes}m`
	return `${minutes}m`
}

/// An RFC 3339 instant rendered as how long ago it was. `last-boot` is shown as an elapsed time
/// (VIEW); a clock the device could not answer for is not sent, so this only runs on a real instant.
function elapsedSince(rfc3339) {
	const then = Date.parse(rfc3339)
	if (Number.isNaN(then)) return String(rfc3339)
	return formatDuration(Math.max(0, (Date.now() - then) / 1000))
}

/// Where a value sits on its scale, from 0 to 1, or null where it has none to be drawn against. Only
/// a fraction, or a quantity with a total or a `limits` trait, has a scale (VIEW).
export function scaleOf(entry, total = null) {
	if (!hasValue(entry)) return null
	if (entry.kind === 'fraction') return clamp(entry.value)
	if (entry.kind === 'quantity') {
		const top = total ?? limitTop(entry)
		if (typeof top === 'number' && top > 0) return clamp(entry.value / top)
	}
	return null
}

/// The largest limit on a reading, which its scale is drawn against where it has no total.
function limitTop(entry) {
	const limits = entry.traits?.limits
	if (!Array.isArray(limits) || limits.length === 0) return null
	return Math.max(...limits.map((limit) => limit.at))
}

function clamp(fraction) {
	return Math.min(1, Math.max(0, fraction))
}

// Series keying. A reading's history is keyed by its catalogue name together with every trait NFO
// does not name as descriptive; an unrecognised trait is part of the key, so a device that gains a
// trait splitting one series into several is left two series it cannot fully tell apart rather than
// one merged wrong one (VIEW).

// The descriptive traits, which do not distinguish one thing measured from another: the whole
// `status`, `limits`, `security` and `channel` traits, and the members named here within the traits
// that hold them. A battery's serial, model and vendor describe the cell; only its name says which
// battery it is. A wireless link's channel moves whenever a shared-channel hotspot follows the client
// onto a new one, and that is the same link, not a second.
const DESCRIPTIVE = new Set(['status', 'limits', 'security', 'channel'])
const DESCRIPTIVE_MEMBERS = {
	interface: new Set(['route', 'overlay']),
	battery: new Set(['serial', 'model', 'vendor']),
}

// The entries NFO's catalogue lists as one entry per value held. Several are held at once with the
// same name, kind and traits, as an interface holds an IPv4 address and more than one IPv6 address,
// so the value is what tells them apart, and one that ends carries it.
const TOLD_APART_BY_VALUE = new Set(['network-address'])

/// The key a reading's history is held under: its name, kind and distinguishing traits, canonicalised
/// so member order does not matter, and its value where that is what tells it apart (NFO).
export function seriesKey(entry) {
	const value = TOLD_APART_BY_VALUE.has(entry.name) ? canonical(entry.value) : ''
	return `${entry.name}\u001f${entry.kind}\u001f${canonical(distinguishing(entry.traits))}\u001f${value}`
}

/// The identity of one entry instance for the tile grid: the same key, so two readings alike but for
/// a distinguishing trait are two entries and the same one re-sampled is one.
export function identityKey(entry) {
	return `${entry.fact ? 'fact' : 'reading'}\u001f${seriesKey(entry)}`
}

/// The traits with the descriptive ones stripped: the whole `status` and `limits`, and the
/// descriptive members of any trait that holds some.
function distinguishing(traits) {
	const kept = {}
	for (const [name, value] of Object.entries(traits ?? {})) {
		if (DESCRIPTIVE.has(name)) continue
		const descriptive = DESCRIPTIVE_MEMBERS[name]
		if (descriptive && value && typeof value === 'object' && !Array.isArray(value)) {
			const trimmed = {}
			for (const [member, inner] of Object.entries(value)) {
				if (!descriptive.has(member)) trimmed[member] = inner
			}
			kept[name] = trimmed
		} else {
			kept[name] = value
		}
	}
	return kept
}

/// A stable string for a JSON value, with object keys sorted, so member order does not change a key.
function canonical(value) {
	if (Array.isArray(value)) return `[${value.map(canonical).join(',')}]`
	if (value && typeof value === 'object') {
		return `{${Object.keys(value)
			.sort()
			.map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`)
			.join(',')}}`
	}
	return JSON.stringify(value)
}

/// Add a reading to its history, keeping the points ordered and trimmed to the window. Facts have no
/// history and are never added (VIEW).
export function pushHistory(history, entry) {
	if (entry.fact || !hasValue(entry)) return history
	const number = numberOf(entry)
	if (number === null) return history
	const key = seriesKey(entry)
	const next = new Map(history)
	const points = [...(next.get(key) ?? []), { at: entry.at, number }]
	points.sort((a, b) => a.at - b.at)
	const newest = points[points.length - 1].at
	next.set(
		key,
		points.filter((point) => newest - point.at <= WINDOW_MS),
	)
	return next
}

/// Drop the history of an entry that has ended, along with its tile (VIEW).
export function forgetHistory(history, entry) {
	const key = seriesKey(entry)
	if (!history.has(key)) return history
	const next = new Map(history)
	next.delete(key)
	return next
}

/// Whether the device has said this entry no longer applies (NFO).
export function isEnded(entry) {
	return statusOf(entry) === 'ended'
}

/// How far back the graph keeps points. A graph fills forward from connection; no history is sent, so
/// this is only the trimming of what accumulates (U1 carries bringing sent history back).
export const WINDOW_MS = 5 * 60 * 1000

/// The number behind a value, where it has one. Null for text and for a kind with no number.
export function numberOf(entry) {
	if (!entry || !hasValue(entry)) return null
	switch (entry.kind) {
		case 'fraction':
		case 'quantity':
			return typeof entry.value === 'number' ? entry.value : null
		case 'duration':
			return typeof entry.value === 'number' ? entry.value : null
		default:
			return null
	}
}

/// Points as an SVG polyline, scaled to their own peak within the box. Each series is scaled to its
/// own peak, and the peak is stated alongside, because two directions of throughput routinely differ
/// by an order of magnitude and a shared scale flattens the quieter one to a line (VIEW).
export function polyline(points, { width, height, flip = false, peak = null }) {
	if (points.length === 0) return { points: '', peak: 0 }
	const top = peak ?? Math.max(...points.map((point) => point.number), 0)
	const first = points[0].at
	const span = points[points.length - 1].at - first || 1

	const coords = points.map((point) => {
		const x = ((point.at - first) / span) * width
		const drawn = top > 0 ? (point.number / top) * height : 0
		const y = flip ? drawn : height - drawn
		return `${round(x)},${round(y)}`
	})
	return { points: coords.join(' '), peak: top }
}

function round(value) {
	return Math.round(value * 10) / 10
}

/// A trait value written for display: a string as-is, an object as its members joined. Used as the
/// qualifier on an entry this build does not otherwise recognise (VIEW).
/// The addresses the Address tile headlines: on the interface carrying the default route, its first
/// IPv4, else global IPv6, else unique local IPv6 address; and on an overlay interface, its first IPv4,
/// else global IPv6 address. At most two, the default route's first (VIEW).
export function headlineAddresses(addresses) {
	const firstOf = (held, classes) => {
		for (const wanted of classes) {
			const found = held.find((entry) => addressClass(entry) === wanted)
			if (found) return found
		}
		return null
	}
	const iface = (entry) => entry.traits?.interface ?? {}
	const onDefault = addresses.filter((entry) => iface(entry).route === 'default' && !iface(entry).overlay)
	const onOverlay = addresses.filter((entry) => Boolean(iface(entry).overlay))
	return [
		firstOf(onDefault, ['ipv4', 'global', 'unique-local']),
		firstOf(onOverlay, ['ipv4', 'global']),
	].filter(Boolean)
}

/// Which of the classes the headline chooses among an address falls in: `ipv4`, `global` IPv6
/// (2000::/3), `unique-local` IPv6 (fc00::/7), or `other`.
function addressClass(entry) {
	if (entry.kind === 'ipv4') return 'ipv4'
	if (entry.kind !== 'ipv6') return 'other'
	const first = parseInt(String(entry.value).split(':')[0] || '0', 16)
	if (first >= 0x2000 && first <= 0x3fff) return 'global'
	if ((first & 0xfe00) === 0xfc00) return 'unique-local'
	return 'other'
}

export function qualifierOf(entry) {
	const parts = []
	for (const [name, value] of Object.entries(entry.traits ?? {})) {
		if (DESCRIPTIVE.has(name)) continue
		parts.push(traitText(value))
	}
	return parts.filter(Boolean).join(' · ')
}

function traitText(value) {
	if (value === null || value === undefined) return ''
	if (typeof value === 'object' && !Array.isArray(value)) {
		return Object.values(value)
			.map((inner) => (typeof inner === 'object' ? JSON.stringify(inner) : String(inner)))
			.join(' ')
	}
	return String(value)
}

File: crates/bliti-core/src/channel/readings.rs
//! One shape for everything a device reports about itself: a fact or a reading.
//!
//! Behaviour is specified in NFO. An [`Entry`] names what it is against a catalogue, says what it is
//! about in its `traits`, and carries a value of some `kind`. A reader renders it from that alone, so
//! a device that gains an entry appears in a reader that has never heard of it with no release in
//! between. A `fact` and a `reading` are the same shape; which catalogue names an entry is the only
//! difference, and it is carried by the message type ([`super::messages`]).
//!
//! `traits` and `value` are kept as raw JSON rather than parsed into fields. That is deliberate:
//! identity is the whole `traits` object, including traits this build cannot read, so a sender that
//! adds a trait splitting one series into several does not have an older reader merge them into one
//! wrong graph. Keeping them raw is also what lets them survive the round trip the envelope's
//! unknown-member detection depends on (see [`super::envelope`]).

use serde_json::{Map, Value as Json};

/// The value kinds this catalogue uses. The vocabulary is open: a reader that meets a kind not here
/// renders the value stringified (NFO, VIEW).
pub mod kind {
	/// A string with no numeric meaning.
	pub const TEXT: &str = "text";
	/// A number from 0 to 1 inclusive, a proportion of a whole.
	pub const FRACTION: &str = "fraction";
	/// A measurement, carrying a unit.
	pub const QUANTITY: &str = "quantity";
	/// An elapsed time, in seconds.
	pub const DURATION: &str = "duration";
	/// An instant, as RFC 3339.
	pub const DATETIME: &str = "datetime";
	/// An IPv4 address.
	pub const IPV4: &str = "ipv4";
	/// An IPv6 address.
	pub const IPV6: &str = "ipv6";
}

/// The `status` trait: how the datum stands. Every entry carries one (NFO).
pub const STATUS: &str = "status";
/// The `limits` trait: marks on a reading's scale.
pub const LIMITS: &str = "limits";

/// The entries NFO's catalogue lists as one entry per value held, which are told apart by their value
/// as well as by their name, kind and distinguishing traits.
pub const TOLD_APART_BY_VALUE: &[&str] = &["network-address"];

/// One fact or reading. One shape; the message type says which catalogue names it.
#[derive(Debug, Clone, PartialEq)]
pub struct Entry {
	/// Milliseconds since the sender booted, when it was taken. Meaningful only against other `at`
	/// values from the same sender.
	pub at: u64,
	/// What it is, named against the catalogue.
	pub name: String,
	/// What it is about: distinguishing traits and descriptive ones, including `status` and any
	/// `limits`. Kept raw so an unrecognised trait survives and still tells two series apart.
	pub traits: Map<String, Json>,
	/// What the value is.
	pub kind: String,
	/// The unit the value is in, named in full, where it has one.
	pub unit: Option<String>,
	/// The value, absent where the status is `skipped` or `broken`.
	pub value: Option<Json>,
}

/// How a datum stands, as the `status` trait's `is` member (NFO).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Status {
	/// The measurement was taken and what it measures is well.
	Passed,
	/// What it measures is degraded, but not gravely.
	Warning,
	/// What it measures is unwell.
	Failed,
	/// A precondition was not met, so nothing was measured.
	Skipped,
	/// The measurement was attempted and errored.
	Broken,
	/// The entry no longer applies, and a reader drops it.
	Ended,
}

impl Status {
	/// The wire string, matching the vocabulary BES software reports checks in.
	pub fn as_str(self) -> &'static str {
		match self {
			Self::Passed => "passed",
			Self::Warning => "warning",
			Self::Failed => "failed",
			Self::Skipped => "skipped",
			Self::Broken => "broken",
			Self::Ended => "ended",
		}
	}

	/// Whether a value accompanies this status.
	fn carries_value(self) -> bool {
		matches!(self, Self::Passed | Self::Warning | Self::Failed)
	}
}

impl Entry {
	/// A passed entry carrying a value. The starting point for everything a device reports; downgrade
	/// it with [`Entry::warning`] and the rest.
	pub fn new(at: u64, name: impl Into<String>, kind: impl Into<String>, value: Json) -> Self {
		let mut entry = Self {
			at,
			name: name.into(),
			traits: Map::new(),
			kind: kind.into(),
			unit: None,
			value: Some(value),
		};
		entry.set_status(Status::Passed, None);
		entry
	}

	/// A fraction, from 0 to 1, rounded to four places.
	pub fn fraction(at: u64, name: impl Into<String>, number: f64) -> Self {
		Self::new(at, name, kind::FRACTION, json_number(round4(number)))
	}

	/// A quantity in a named unit, rounded to four places. The unit is spelled out in full (NFO).
	pub fn quantity(
		at: u64,
		name: impl Into<String>,
		unit: impl Into<String>,
		number: f64,
	) -> Self {
		let mut entry = Self::new(at, name, kind::QUANTITY, json_number(round4(number)));
		entry.unit = Some(unit.into());
		entry
	}

	/// An elapsed time, in seconds.
	pub fn duration(at: u64, name: impl Into<String>, seconds: f64) -> Self {
		Self::new(at, name, kind::DURATION, json_number(round4(seconds)))
	}

	/// Text with no numeric meaning.
	pub fn text(at: u64, name: impl Into<String>, text: impl Into<String>) -> Self {
		Self::new(at, name, kind::TEXT, Json::String(text.into()))
	}

	/// An instant, as RFC 3339.
	pub fn datetime(at: u64, name: impl Into<String>, rfc3339: impl Into<String>) -> Self {
		Self::new(at, name, kind::DATETIME, Json::String(rfc3339.into()))
	}

	/// An internet address, of the kind its family names.
	pub fn address(
		at: u64,
		name: impl Into<String>,
		kind: impl Into<String>,
		ip: impl Into<String>,
	) -> Self {
		Self::new(at, name, kind, Json::String(ip.into()))
	}

	/// Give a trait a value. A trait that only qualifies another belongs inside it, not beside it, so
	/// pass an object where a trait has more than one thing to say (NFO).
	#[must_use]
	pub fn with_trait(mut self, name: impl Into<String>, value: Json) -> Self {
		self.traits.insert(name.into(), value);
		self
	}

	/// Add a mark to the reading's scale. Kept as the `limits` trait, a list of `{at, label}`.
	#[must_use]
	pub fn with_limit(mut self, at: f64, label: impl Into<String>) -> Self {
		let mark = serde_json::json!({ "at": round4(at), "label": label.into() });
		match self.traits.get_mut(LIMITS) {
			Some(Json::Array(marks)) => marks.push(mark),
			_ => {
				self.traits
					.insert(LIMITS.to_owned(), Json::Array(vec![mark]));
			}
		}
		self
	}

	/// Report the entry degraded, keeping its value. `reason` is the sender's own words (NFO).
	#[must_use]
	pub fn warning(mut self, reason: impl Into<String>) -> Self {
		self.set_status(Status::Warning, Some(reason.into()));
		self
	}

	/// Report what the entry measures as unwell, keeping its value.
	#[must_use]
	pub fn failed(mut self, reason: impl Into<String>) -> Self {
		self.set_status(Status::Failed, Some(reason.into()));
		self
	}

	/// An entry whose precondition was not met: no value, with a reason.
	pub fn skipped(
		at: u64,
		name: impl Into<String>,
		kind: impl Into<String>,
		reason: impl Into<String>,
	) -> Self {
		let mut entry = Self {
			at,
			name: name.into(),
			traits: Map::new(),
			kind: kind.into(),
			unit: None,
			value: None,
		};
		entry.set_status(Status::Skipped, Some(reason.into()));
		entry
	}

	/// An entry the device declared but could not take: no value, with a reason. Hardware that is not
	/// fitted is left out entirely; this is for hardware that is there and did not answer (NFO).
	pub fn broken(
		at: u64,
		name: impl Into<String>,
		kind: impl Into<String>,
		reason: impl Into<String>,
	) -> Self {
		let mut entry = Self {
			at,
			name: name.into(),
			traits: Map::new(),
			kind: kind.into(),
			unit: None,
			value: None,
		};
		entry.set_status(Status::Broken, Some(reason.into()));
		entry
	}

	/// This entry, sent once more to say it no longer applies: everything that told it apart, so a
	/// reader knows which entry it drops, and the reason it ended. Only an entry told apart by its
	/// value keeps the value, as the one that no longer holds (NFO).
	pub fn ended(&self, at: u64, reason: impl Into<String>) -> Self {
		let mut entry = self.clone();
		entry.at = at;
		entry.set_status(Status::Ended, Some(reason.into()));
		if self.told_apart_by_value() {
			entry.value = self.value.clone();
		}
		entry
	}

	/// Whether this entry is told apart by its value, being one of several held at once (NFO).
	pub fn told_apart_by_value(&self) -> bool {
		TOLD_APART_BY_VALUE.contains(&self.name.as_str())
	}

	/// Set the `status` trait, tying value presence to it: present for `passed`, `warning` and
	/// `failed`, absent for `skipped`, `broken` and `ended` (NFO).
	fn set_status(&mut self, status: Status, reason: Option<String>) {
		let mut object = Map::new();
		object.insert("is".to_owned(), Json::String(status.as_str().to_owned()));
		if let Some(reason) = reason {
			object.insert("reason".to_owned(), Json::String(reason));
		}
		self.traits.insert(STATUS.to_owned(), Json::Object(object));
		if !status.carries_value() {
			self.value = None;
		}
	}

	/// This entry's status, as its `is` string. A status this build does not know is kept as it
	/// arrived, so an older reader colours it as if passed rather than crying wolf.
	pub fn status(&self) -> Option<&str> {
		self.traits.get(STATUS)?.get("is")?.as_str()
	}

	/// The reason on the status, where there is one.
	pub fn reason(&self) -> Option<&str> {
		self.traits.get(STATUS)?.get("reason")?.as_str()
	}

	/// Write this entry into a message map, naming it against the given member (`fact` or
	/// `measurement`). No member it holds is skipped, so the round trip the envelope depends on stays
	/// lossless.
	pub(super) fn write_into(&self, name_member: &str, map: &mut Map<String, Json>) {
		map.insert("at".to_owned(), Json::Number(self.at.into()));
		map.insert(name_member.to_owned(), Json::String(self.name.clone()));
		map.insert("traits".to_owned(), Json::Object(self.traits.clone()));
		map.insert("kind".to_owned(), Json::String(self.kind.clone()));
		if let Some(unit) = &self.unit {
			map.insert("unit".to_owned(), Json::String(unit.clone()));
		}
		if let Some(value) = &self.value {
			map.insert("value".to_owned(), value.clone());
		}
	}

	/// Read an entry from a message map, taking its name from the given member. The members a fact and
	/// a reading required when they were defined are checked here; a missing one is a malformed
	/// message rather than a newer peer (MSG).
	pub(super) fn read_from(name_member: &str, map: &Map<String, Json>) -> Result<Self, String> {
		let at = map
			.get("at")
			.and_then(Json::as_u64)
			.ok_or("an entry carries a number `at`")?;
		let name = map
			.get(name_member)
			.and_then(Json::as_str)
			.ok_or_else(|| format!("an entry carries a string `{name_member}`"))?
			.to_owned();
		let traits = match map.get("traits") {
			Some(Json::Object(traits)) => traits.clone(),
			_ => return Err("an entry carries an object `traits`".to_owned()),
		};
		let kind = map
			.get("kind")
			.and_then(Json::as_str)
			.ok_or("an entry carries a string `kind`")?
			.to_owned();
		let unit = match map.get("unit") {
			None | Some(Json::Null) => None,
			Some(Json::String(unit)) => Some(unit.clone()),
			Some(_) => return Err("`unit` is a string".to_owned()),
		};
		let value = map.get("value").cloned();
		Ok(Self {
			at,
			name,
			traits,
			kind,
			unit,
			value,
		})
	}
}

/// A JSON number, falling back to null for a value JSON cannot carry.
fn json_number(value: f64) -> Json {
	serde_json::Number::from_f64(value).map_or(Json::Null, Json::Number)
}

/// Round to at most four decimal places. No reading this protocol carries is meaningful past four,
/// and rounding nearly halves what a stream of them costs compressed (NFO).
fn round4(value: f64) -> f64 {
	(value * 10_000.0).round() / 10_000.0
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn a_passed_entry_carries_its_value_and_a_status() {
		let entry = Entry::fraction(20_308_140, "cpu-usage", 0.1234);
		assert_eq!(entry.status(), Some("passed"));
		assert!(entry.reason().is_none(), "passed carries no reason");
		assert_eq!(entry.value, Some(json_number(0.1234)));
	}

	/// An address that goes keeps its value, the only thing telling it from the interface's others,
	/// and any other entry that ends loses its value.
	#[test]
	fn an_ended_entry_keeps_its_value_only_where_the_value_tells_it_apart() {
		let address = Entry::address(1, "network-address", kind::IPV6, "fd00::1".to_owned());
		let ended = address.ended(2, "no longer applies");
		assert_eq!(ended.status(), Some("ended"));
		assert_eq!(ended.value, Some(Json::String("fd00::1".into())));

		let hotspot = Entry::text(1, "hotspot", "bliti");
		assert_eq!(hotspot.ended(2, "no longer applies").value, None);
	}

	#[test]
	fn a_numeric_value_is_rounded_to_four_places() {
		let entry = Entry::fraction(1, "cpu-usage", 0.123_456_789);
		assert_eq!(entry.value, Some(json_number(0.1235)));
		let quantity = Entry::quantity(1, "temperature", "celsius", 48.567_89);
		assert_eq!(quantity.value, Some(json_number(48.5679)));
	}

	#[test]
	fn a_warning_keeps_its_value_and_carries_a_reason() {
		let entry = Entry::quantity(1, "cpu-frequency", "hertz", 600_000_000.0)
			.warning("the platform is limiting the processor");
		assert_eq!(entry.status(), Some("warning"));
		assert_eq!(
			entry.reason(),
			Some("the platform is limiting the processor")
		);
		assert!(entry.value.is_some());
	}

	#[test]
	fn skipped_and_broken_carry_no_value_but_a_reason() {
		let skipped = Entry::skipped(
			1,
			"battery-direction",
			kind::TEXT,
			"not watched long enough yet",
		);
		assert_eq!(skipped.status(), Some("skipped"));
		assert_eq!(skipped.reason(), Some("not watched long enough yet"));
		assert!(skipped.value.is_none());

		let broken = Entry::broken(
			1,
			"temperature",
			kind::QUANTITY,
			"no answer from the sensor",
		);
		assert_eq!(broken.status(), Some("broken"));
		assert!(broken.value.is_none());
	}

	#[test]
	fn an_ended_entry_keeps_its_traits_and_drops_its_value() {
		let running = Entry::new(1, "hotspot", "text", serde_json::json!("clinic"))
			.with_trait("channel", serde_json::json!({"number": 6}));
		let ended = running.ended(2, "the hotspot stopped");
		assert_eq!(ended.status(), Some("ended"));
		assert_eq!(ended.reason(), Some("the hotspot stopped"));
		assert_eq!(ended.value, None);
		assert_eq!(ended.at, 2);
		assert_eq!(ended.traits.get("channel"), running.traits.get("channel"));
	}

	#[test]
	fn a_trait_that_qualifies_another_sits_inside_it() {
		let entry = Entry::address(1, "network-address", kind::IPV4, "192.168.1.42").with_trait(
			"interface",
			serde_json::json!({ "name": "eth0", "route": "default" }),
		);
		let interface = entry.traits.get("interface").unwrap();
		assert_eq!(
			interface.get("route").and_then(Json::as_str),
			Some("default")
		);
	}

	#[test]
	fn an_entry_round_trips_through_a_map() {
		let entry = Entry::quantity(
			20_308_140,
			"network-throughput",
			"bytes/second",
			1_200_000.0,
		)
		.with_trait("interface", serde_json::json!({ "name": "eth0" }))
		.with_trait("direction", Json::String("in".to_owned()));

		let mut map = Map::new();
		entry.write_into("measurement", &mut map);
		assert_eq!(
			map.get("measurement").and_then(Json::as_str),
			Some("network-throughput")
		);
		assert!(map.get("value").is_some());

		let read = Entry::read_from("measurement", &map).unwrap();
		assert_eq!(read, entry);
	}

	#[test]
	fn reading_an_entry_missing_what_it_requires_is_an_error() {
		let mut map = Map::new();
		map.insert(
			"measurement".to_owned(),
			Json::String("cpu-usage".to_owned()),
		);
		// No `at`, no `traits`, no `kind`.
		assert!(Entry::read_from("measurement", &map).is_err());
	}
}
